tools/tesh/tesh.1
.\" Manpage for tesh, the TEsting SHell.
.\" 
.\" To read it locally (when not installed): 
.\"   man ./tesh.1
.\"
.TH tesh 1 "10 Oct 2012" "1.0" "tesh man page"
.SH NAME
tesh \- testing shell
.SH SYNOPSIS
.B tesh
[\fIOPTION\fR]... [\fIFILE\fR]...
.SH DESCRIPTION
This is the TESH tool. It constitutes a testing shell, ie a sort of shell specialized to run tests. The list of actions to take is parsed from files files called testsuite.
.SH OPTIONS
  --cd some/directory : ask tesh to switch the working directory before
                        launching the tests
  --setenv var=value  : set a specific environment variable
  --cfg arg           : add parameter --cfg=arg to each command line
  --enable-coverage   : ignore output lines starting with "profiling:"
.SH TESH FILE SYNTAX
Here is the syntax of these files:

The kind of each line is given by the first char (the second char should be
blank and is ignored):

 `$' command to run in foreground
 `&' command to run in background
 `<' input to pass to the command
 `>' output expected from the command
 `!' metacommand, which can be one of:
     `timeout' <integer>|no
     `expect signal' <signal name>
     `expect return' <integer>
     `output' <ignore|display>
     `setenv <key>=<val>'
 `p' a string to print
 `P' a string to print at the CRITICAL level (ease logging grepping)

If the expected output do not match what the command spits, TESH will produce
an error showing the diff (see OUTPUT below).
.SH IO ORDERS
The < and > lines add IO to the command defined in the current block (blocks
are separated by blank lines). It is possible to place these lines either after
the command or before. The difference between the two following chunks is
mainly cosmetic in your testsuites, TESH don't care. (cf IO-orders.tesh)

 $ cat
 < TOTO
 > TOTO

 > TOTO
 $ cat
 < TOTO

Nevertheless, it is possible to have several commands in the same block, but
none of them can have any output. It may seem a bit restrictive, as one could
say that a command gets all the IO until the next command, but I'm afraid of
errors such as the following:

 $ cd toto
 > TOTO
 $ mkfile file

TOTO will be passed to the cd command, where the user clearly want to pass it
to the mkfile built-in command (see below).
.SH STREAM REDIRECTION
Stream redirections (">", "<" and "|" constructs in sh) are not
implemented yet in tesh. This is a bit restrictive, but well, patch
welcome...

The situation in which it is mainly problematic is to create a
temporary file. The solution is to use the "mkfile" built-in command,
as in the following example:
$ mkfile myFile
> some content
> to the file

This will create a file called myFile (first argument of the mkfile
command). Its content will be all the input provided to the command.
.SH RETURN CODE
TESH spits an appropriate error message when the child do not return 0 as
return code (cf. catch-return.tesh), and returns code+40 itself.

It is also possible to specify that a given command must return another
value. For this, use the "expect return" metacommand, which takes an integer as
argument. The change only apply to the next command (cf. set-return.tesh).
.SH SIGNALS
TESH detects when the child is killed by a signal (like on segfaults), and
spits an appropriate error message (cf. catch-signal.tesh).

It is also possible to specify that a given command must raise a given
signal. For this, use the "expect signal" metacommand. It takes the signal name
as argument. The change only apply to the next command (cf. set-signal.tesh).
.SH TIMEOUTS
By default, all commands are given 5 seconds to execute
(cf. catch-timeout.tesh). You can change this with the "timeout", which
takes an integer as argument. The change only apply to the next command
(cf. set-timeout.tesh). If you pass "no" as argument, the command
cannot timeout.
.SH OUTPUT
By default, the commands output is matched against the one expected,
and an error is raised on discrepancy. Metacommands to change this:
 "output ignore"  -> output completely discarded
 "output display" -> output displayed (but not verified)
 "output sort"    -> sorts the display before verifying it (see below)
.SH SORTING OUTPUT

SimGrid is designed to produce perfectly reproducible results, so its
output can usualy be compared without any preprocessing. This is not
true anymore when the user activates parallel execution: User
processes are run in parallel at each timestamp, and the output is not
reproducible anymore. Until you sort the lines.

If you ask for 
.B ! output sort
then tesh will sort the whole lines. But it really complicates the
analysis of the error detected: the logical order of the output is
defeated by the lexicographical sort. 

The solution is to ask for
.B ! output sort 19
instead to sort only on the prefix of the line. Indeed, we run our simulation
tests with the flag: 
  --log=root.fmt:[%10.6r]%e(%i:%P@%h)%e%m%n

Then, the previous command sorts  lines on the first 19 chars, that is
exactly the length of the prefix indicating the timestamp and the
process. That's exactly what we need:
 - Every timestamps remain separated, as it should; 
 - In each timestamp, the output order of processes become
   reproducible: that's the lexicographical order of their name;
 - For each process, the order of its execution is preserved: its
   messages within a given timestamp are not reordered.

That way, tesh can do its job (no false positive, no false negative)
despite the unpredictable order of executions of processes within a
timestamp, and reported errors remain easy to analyze (execution of a
given process preserved).

This is of course very SimGrid oriented, but could even be usable by
others, who knows?

.SH ENVIRONMENT
You can add some content to the tested processes environment with the
setenv metacommand. It works as expected. For example:
  "setenv PATH=/bin"
.SH BUGS
No known bugs. 


NEWS
                    _               _____  _ ____
__   _____ _ __ ___(_) ___  _ __   |___ / / |___ \
\ \ / / _ \ '__/ __| |/ _ \| '_ \    |_ \ | | __) |
 \ V /  __/ |  \__ \ | (_) | | | |  ___) || |/ __/
  \_/ \___|_|  |___/_|\___/|_| |_| |____(_)_|_____|

The Facelift Release. Major changes:

 * Many interface glitches addressed, everywhere.
   - Require g++-4.7 and boost 1.48
   - See ChangeLog for renamed functions and options.

 * Energy plugin: major cleanups/rewrites.
   http://simgrid.org/tutorials/simgrid-energy-101.pdf

 * Model-Checker progresses toward production use:
   - More robust (now runs in a separate process).
   - More efficient (with snapshots' memory compaction).
   - More usable (execution paths replayed out of MC)
   http://simgrid.org/tutorials/simgrid-mc-101.pdf

 * SMPI improvements:
   - Better standard coverage  - SMP-aware collectives
   - InfiniBand model          - Early support for RMA
   - Replay: Dynamic selector  - MVAPICH2 selector
   http://simgrid.org/tutorials/simgrid-smpi-101.pdf

Upcoming v3.13 will introduce new interfaces, toward SimGrid 4.

                    _               _____  _ _
__   _____ _ __ ___(_) ___  _ __   |___ / / / |
\ \ / / _ \ '__/ __| |/ _ \| '_ \    |_ \ | | |
 \ V /  __/ |  \__ \ | (_) | | | |  ___) || | |
  \_/ \___|_|  |___/_|\___/|_| |_| |____(_)_|_|
                May 31 2014

The Class Release. Major changes:

 * Surf is now in C++ (and documented!)
 * Virtual Machine model for Cloud Simulation
 * Surf callbacks: plug your code directly in Surf!
 * Simcalls are script-generated to clean the mess

                    _               _____  _  ___
__   _____ _ __ ___(_) ___  _ __   |___ / / |/ _ \
\ \ / / _ \ '__/ __| |/ _ \| '_ \    |_ \ | | | | |
 \ V /  __/ |  \__ \ | (_) | | | |  ___) || | |_| |
  \_/ \___|_|  |___/_|\___/|_| |_| |____(_)_|\___/
                Nov 17 2013

The Clean Diaper Release, a.k.a. SimGrid is leak-free. Major changes:

 * Preliminary DVFS support to track the energy consumption
 * Java is back in the main package (and is very stable).
 * The storage sub-modules is now believed to be usable.
 * SMPI is now very stable (we pass most MPICH tests).
 * Lots of memory leaks were corrected in this release.
 * Verification and model checking further improved:
   liveness+SMPI works; reduction through state equality detection
 Plus the usual load of bug fixes and small improvements

                    _               _____ ___
__   _____ _ __ ___(_) ___  _ __   |___ // _ \
\ \ / / _ \ '__/ __| |/ _ \| '_ \    |_ \ (_) |
 \ V /  __/ |  \__ \ | (_) | | | |  ___) \__, |
  \_/ \___|_|  |___/_|\___/|_| |_| |____(_)/_/
                Feb 5 2013

The "Grasgory" release. Major changes:

 * Gras was completely removed from this version.
 * Documentation reorganization to ease browsing it.
 * New default value for the TCP_gamma parameter: 4MiB

                    _               _____  ___   _
__   _____ _ __ ___(_) ___  _ __   |___ / ( _ ) / |
\ \ / / _ \ '__/ __| |/ _ \| '_ \    |_ \ / _ \ | |
 \ V /  __/ |  \__ \ | (_) | | | |  ___) | (_) || |
  \_/ \___|_|  |___/_|\___/|_| |_| |____(_)___(_)_|
                Oct 27 2012

The "we are told that some people want to also *install* the simgrid
framework" release.

 * Add missing manpage to the archive.

                    _               _____  ___
__   _____ _ __ ___(_) ___  _ __   |___ / ( _ )
\ \ / / _ \ '__/ __| |/ _ \| '_ \    |_ \ / _ \
 \ V /  __/ |  \__ \ | (_) | | | |  ___) | (_) |
  \_/ \___|_|  |___/_|\___/|_| |_| |____(_)___/
                Oct 25 2012

The Psssshiiiit release: SimGrid jumps into the Cloud.
Major changes:

 * Experimental interface to manipulate VMs, EC2-style.
 * Fixes around process restart and stochastic workloads
 * platf: New C interface to create fixed or random platforms
 * SimDag: Many fixes and improvements of typed tasks
 * SMPI now covers more of the MPI interface.
   More datatypes, more functions, more robust.
 * Model-checking: mmalloc is more robust to user errors.

                    _               _____ _____ _
__   _____ _ __ ___(_) ___  _ __   |___ /|___  / |
\ \ / / _ \ '__/ __| |/ _ \| '_ \    |_ \   / /| |
 \ V /  __/ |  \__ \ | (_) | | | |  ___) | / /_| |
  \_/ \___|_|  |___/_|\___/|_| |_| |____(_)_/(_)_|
                June 7 2012

The "cleaning the cleanup" release. Major changes:

 * Portability fixups for Mac OSX and Windows
 * Some other bug fixing and various polishing.

                    _               ____  _____
__   _____ _ __ ___(_) ___  _ __   |___ /|___  |
\ \ / / _ \ '__/ __| |/ _ \| '_ \    |_ \   / /
 \ V /  __/ |  \__ \ | (_) | | | |  ___) | / /
  \_/ \___|_|  |___/_|\___/|_| |_| |____(_)_/
                May 15 2012

The "spring cleanups (before the next Big Project kicks in)" release.
Major changes:

 * Major cleanups all around (doc, user options, MSG, Lua, internals)
   Small backward compatibility glitches may have been introduced
 * Parallel execution of user code should be stable and efficient now
 * SMPI is now considered stable
 * Add temporals to Model-Checking (highly experimental right now)

                    _               _____  __    ____
__   _____ _ __ ___(_) ___  _ __   |___ / / /_  |___ \
\ \ / / _ \ '__/ __| |/ _ \| '_ \    |_ \| '_ \   __) |
 \ V /  __/ |  \__ \ | (_) | | | |  ___) | (_) | / __/
  \_/ \___|_|  |___/_|\___/|_| |_| |____(_)___(_)_____|
                Oct 5 2011

The "Not coding new stuff allows to polish old things" release.

 * Portability to Mac and Windows improved.
 * Possible misconfigurations (contexts, libPCRE) made impossible by
   removing the option or providing sane default value.
 * Experimental support to NS3 simulator as a back-end.

                    _               _____  __    _
__   _____ _ __ ___(_) ___  _ __   |___ / / /_  / |
\ \ / / _ \ '__/ __| |/ _ \| '_ \    |_ \| '_ \ | |
 \ V /  __/ |  \__ \ | (_) | | | |  ___) | (_) || |
  \_/ \___|_|  |___/_|\___/|_| |_| |____(_)___(_)_|
                Jun 27 2011


tools/tesh/tesh.pl
#! /usr/bin/perl

# Copyright (c) 2012-2014. The SimGrid Team.
# All rights reserved.

# This program is free software; you can redistribute it and/or modify it
# under the terms of the license (GNU LGPL) which comes with this package.
eval 'exec perl -S $0 ${1+"$@"}'
  if $running_under_some_shell;

=encoding UTF-8

=head1 NAME

tesh -- testing shell

=head1 SYNOPSIS

B<tesh> [I<options>] I<tesh_file>

=cut
my($bindir)=".";
my($srcdir)=".";
my($timeout)=0;
my($time_to_wait)=0;
my $path = $0;
my $OS;
my $enable_coverage=0;
my $diff_tool=0;
my $diff_tool_tmp_fh=0;
my $diff_tool_tmp_filename=0;
my $sort_prefix = -1;
my $tesh_file;
my $tesh_name;
my $error=0;
my $exitcode=0;
my @bg_cmds;
my (%environ);
$SIG{'PIPE'} = 'IGNORE';
$path =~ s|[^/]*$||;
push @INC,$path;

use Getopt::Long qw(GetOptions);
use strict;
use Term::ANSIColor;
use Text::ParseWords;
use IPC::Open3;
use IO::File;

if($^O eq "linux"){
    $OS = "UNIX";
}
else{
    $OS = "WIN";
    $ENV{"PRINTF_EXPONENT_DIGITS"} = "2";
}

##
## Command line option handling
##

sub var_subst {
    my ($text, $name, $value) = @_;
    if ($value) {
        $text =~ s/\${$name(?::[=-][^}]*)?}/$value/g;
        $text =~ s/\$$name(\W|$)/$value$1/g;
    }
    else {
        $text =~ s/\${$name:=([^}]*)}/$1/g;
        $text =~ s/\${$name}//g;
        $text =~ s/\$$name(\W|$)/$1/g;
    }
    return $text;
}

# option handling helper subs
sub cd_cmd {
  my $directory=$_[1];
  my $failure=1;
  if (-e $directory && -d $directory) {
    chdir("$directory");
    print "[Tesh/INFO] change directory to $directory\n";
  $failure=0;
  } elsif (-e $directory) {
    print "Cannot change directory to '$directory': it is not a directory\n";
  } else {
    print "Chdir to $directory failed: No such file or directory\n";
  }
  if($failure==1){
  $error=1;
  $exitcode=4;
  print "Test suite `$tesh_file': NOK (system error)\n";
  exit 4;
  }
}

sub setenv_cmd {
  my($var,$ctn);
  if ($_[0] =~ /^(.*)=(.*)$/) {
    ($var,$ctn)=($1,$2);
  }elsif ($_[1] =~ /^(.*)=(.*)$/) {
    ($var,$ctn)=($1,$2);
  } else {
      die "[Tesh/CRITICAL] Malformed argument to setenv: expected 'name=value' but got '$_[1]'\n";
  }

  print "[Tesh/INFO] setenv $var=$ctn\n";
  $environ{$var} = $ctn;
}

# Main option parsing sub

sub get_options {
  # remove the tesh file from the ARGV used
  my @ARGV = @_;
  $tesh_file = pop @ARGV;

  # temporary arrays for GetOption
  my @verbose = ();
  my @cfg;
  my $log; # ignored


  my %opt = (
    "help"  => 0,
    "debug"   => 0,
    "verbose" => 0
    );

  Getopt::Long::config('bundling', 'no_getopt_compat', 'no_auto_abbrev');

  GetOptions(
    'help|h'   => \$opt{'help'},

    'verbose|v'  => \@verbose,
    'debug|d'  => \$opt{"debug"},

    'difftool=s' => \$diff_tool,

    'cd=s'     => \&cd_cmd,
    'timeout=s'  => \$opt{'timeout'},
    'setenv=s'   => \&setenv_cmd,
    'cfg=s'    => \@cfg,
    'log=s'    => \$log,
    'enable-coverage+'  => \$enable_coverage,
    );

  if($enable_coverage){
    print "Enable coverage\n";
  }

  if($diff_tool){
    use File::Temp qw/ tempfile /;
    ($diff_tool_tmp_fh, $diff_tool_tmp_filename) = tempfile();
    print "New tesh: $diff_tool_tmp_filename\n";
  }

  unless($tesh_file=~/(.*)\.tesh/){
    $tesh_file="(stdin)";
    $tesh_name="(stdin)";
    print "Test suite from stdin\n";
  }else{
    $tesh_name=$1;
    print "Test suite `$tesh_name'\n";
  }

  $opt{'verbose'} = scalar @verbose;
  foreach (@cfg) {
    $opt{'cfg'} .= " --cfg=$_";
  }
  return %opt;
}

my %opts = get_options(@ARGV);

##
## File parsing
##
my($nb_arg)=0;
my($old_buffer);
my($linebis);
my($SIGABRT)=0;
my($verbose)=0;
my($return)=-1;
my($pid);
my($result);
my($result_err);
my($forked);
my($config)="";
my($tesh_command)=0;
my(@buffer_tesh)=();

#eval {
  use POSIX;

  sub exit_status {
    my $status = shift;
    if (POSIX::WIFEXITED($status)) {
      $exitcode=POSIX::WEXITSTATUS($status)+40;
      return "returned code ".POSIX::WEXITSTATUS($status);
    } elsif (POSIX::WIFSIGNALED($status)) {
      my $code;
      if (POSIX::WTERMSIG($status) == SIGINT){$code="SIGINT"; }
      elsif  (POSIX::WTERMSIG($status) == SIGTERM) {$code="SIGTERM"; }
      elsif  (POSIX::WTERMSIG($status) == SIGKILL) {$code= "SIGKILL"; }
      elsif  (POSIX::WTERMSIG($status) == SIGABRT) {$code="SIGABRT"; }
      elsif  (POSIX::WTERMSIG($status) == SIGSEGV) {$code="SIGSEGV" ;}
      $exitcode=POSIX::WTERMSIG($status)+4;
      return "got signal $code";
    }
    return "Unparsable status. Is the process stopped?";
  }
#};
#if ($@) { # no POSIX available?
#  warn "POSIX not usable to parse the return value of forked child: $@\n";
#  sub exit_status {
#    return "returned code -1 $@ ";
#  }
#}

sub exec_cmd {
  my %cmd = %{$_[0]};
  if ($opts{'debug'}) {
    print "IN BEGIN\n";
    map {print "  $_"} @{$cmd{'in'}};
    print "IN END\n";
    print "OUT BEGIN\n";
    map {print "  $_"} @{$cmd{'out'}};
    print "OUT END\n";
    print "CMD: $cmd{'cmd'}\n";
  }

  # cleanup the command line
  if($OS eq "WIN") {
      var_subst($cmd{'cmd'}, "EXEEXT", ".exe");
  } else {
      var_subst($cmd{'cmd'}, "EXEEXT", "");
  }

  # substitute environ variables
  foreach my $key (keys %environ) {
      $cmd{'cmd'} = var_subst($cmd{'cmd'}, $key, $environ{$key});
  }
  # substitute remaining variables, if any
  while ($cmd{'cmd'} =~ /\${(\w+)(?::[=-][^}]*)?}/) {
      $cmd{'cmd'} = var_subst($cmd{'cmd'}, $1, "");
  }
  while ($cmd{'cmd'} =~ /\$(\w+)/) {
      $cmd{'cmd'} = var_subst($cmd{'cmd'}, $1, "");
  }

  # add cfg options
  $cmd{'cmd'} .= " $opts{'cfg'}" if (defined($opts{'cfg'}) && length($opts{'cfg'}));

  # final cleanup
  $cmd{'cmd'} =~ s/^\s+//;
  $cmd{'cmd'} =~ s/\s+$//;

  print "[$tesh_name:$cmd{'line'}] $cmd{'cmd'}\n" ;

  ###
  # exec the command line
  ###  $line =~ s/\r//g;

  $cmd{'got'} = IO::File->new_tmpfile;
  $cmd{'got'}->autoflush(1);
  local *E = $cmd{'got'};
  $cmd{'pid'} = open3(\*CHILD_IN,  ">&E",  ">&E",
                      quotewords('\s+', 0, $cmd{'cmd'}));

  # push all provided input to executing child
  map { print CHILD_IN "$_\n"; }  @{$cmd{'in'}};
  close CHILD_IN;

  # if timeout specified, fork and kill executing child at the end of timeout
  if (defined($cmd{'timeout'}) or defined($opts{'timeout'})){
    $time_to_wait= defined($cmd{'timeout'}) ? $cmd{'timeout'} : $opts{'timeout'};
    $forked = fork();
    $timeout=-1;
    die "fork() failed: $!" unless defined $forked;
    if ( $forked == 0 ) { # child
      sleep $time_to_wait;
      kill(SIGKILL, $cmd{'pid'});
      exit $time_to_wait;
    }
  }


  # Cleanup the executing child, and kill the timeouter brother on need
  $cmd{'return'} = 0 unless defined($cmd{'return'});
  if($cmd{'background'} != 1){
    waitpid ($cmd{'pid'}, 0);
    $cmd{'gotret'} = exit_status($?);
    parse_out(\%cmd);
  }else{
    # & commands, which will be handled at the end
    push @bg_cmds, \%cmd;
    # no timeout for background commands
    if($forked){
       kill(SIGKILL, $forked);
       $timeout=0;
       $forked=0;
    }
  }
}


sub parse_out {
  my %cmd = %{$_[0]};
  my $gotret=$cmd{'gotret'};

  my $wantret;

  if(defined($cmd{'expect'}) and ($cmd{'expect'} ne "")){
    $wantret = "got signal $cmd{'expect'}";
  }else{
    $wantret = "returned code ".(defined($cmd{'return'})? $cmd{'return'} : 0);
  }

  local *got = $cmd{'got'};
  seek(got,0,0);
  # pop all output from executing child
  my @got;
  while(defined(my $got=<got>)) {
    $got =~ s/\r//g;
    chomp $got;
    print $diff_tool_tmp_fh "> $got\n" if ($diff_tool);

    if (!($enable_coverage and $got=~ /^profiling:/)){
      push @got, $got;
    }
  }

  if ($cmd{'sort'}){
    # Save the unsorted observed output to report it on error.
    map { push @{$cmd{'unsorted got'}}, $_ } @got;

    sub mysort{
        substr($a, 0, $sort_prefix) cmp substr($b, 0, $sort_prefix)
    }
    use sort 'stable';
    if ($sort_prefix>0) {
	@got = sort mysort @got;
    } else {
	@got = sort @got;
    }	    
    while (@got and $got[0] eq "") {
      shift @got;
    }

    # Sort the expected output to make it easier to write for humans
    if(defined($cmd{'out'})){
      if ($sort_prefix>0) {
	  @{$cmd{'out'}} = sort mysort @{$cmd{'out'}};
      } else {
	  @{$cmd{'out'}} = sort @{$cmd{'out'}};
      }
      while (@{$cmd{'out'}} and ${$cmd{'out'}}[0] eq "") {
        shift @{$cmd{'out'}};
      }
    }
  }

  # Did we timeout ? If yes, handle it. If not, kill the forked process.

  if($timeout==-1 and $gotret eq "got signal SIGKILL"){
    $gotret="return code 0";
    $timeout=1;
    $gotret= "timeout after $time_to_wait sec";
    $error=1;
    $exitcode=3;
    print STDERR "<$cmd{'file'}:$cmd{'line'}> timeouted. Kill the process.\n";
  }else{
    $timeout=0;
  }
  if($gotret ne $wantret) {
    $error=1;
    my $msg = "Test suite `$cmd{'file'}': NOK (<$cmd{'file'}:$cmd{'line'}> $gotret)\n";
    if ($timeout!=1) {
        $msg=$msg."Output of <$cmd{'file'}:$cmd{'line'}> so far:\n";
    }
    map {$msg .=  "|| $_\n"} @got;
    if(!@got) {
        if($timeout==1){
        print STDERR "<$cmd{'file'}:$cmd{'line'}> No output before timeout\n";
        }else{
        $msg .= "||\n";
        }
    }
    $timeout = 0;
    print STDERR "$msg";
  }


  ###
  # Check the result of execution
  ###
  my $diff;
  if (defined($cmd{'output display'})){
    print "[Tesh/INFO] Here is the (ignored) command output:\n";
    map { print "||$_\n" } @got;
  }
  elsif (!defined($cmd{'output ignore'})){
    $diff = build_diff(\@{$cmd{'out'}}, \@got);
  }else{
    print "(ignoring the output of <$cmd{'file'}:$cmd{'line'}> as requested)\n"
  }
  if (length $diff) {
    print "Output of <$cmd{'file'}:$cmd{'line'}> mismatch".($cmd{'sort'}?" (even after sorting)":"").":\n";
    map { print "$_\n" } split(/\n/,$diff);
    if ($cmd{'sort'}) {
	print "WARNING: Both the observed output and expected output were sorted as requested.\n";
	print "WARNING: Output were only sorted using the $sort_prefix first chars.\n"
	  if ($sort_prefix>0);
	print "WARNING: Use <! output sort 19> to sort by simulated date and process ID only.\n";
	# print "----8<---------------  Begin of unprocessed observed output (as it should appear in file):\n";
	# map {print "> $_\n"} @{$cmd{'unsorted got'}};
	# print "--------------->8----  End of the unprocessed observed output.\n";
    }

    print "Test suite `$cmd{'file'}': NOK (<$cmd{'file'}:$cmd{'line'}> output mismatch)\n";
    $error=1;
    $exitcode=2;
  }
}

sub mkfile_cmd {
  my %cmd = %{$_[0]};
  my $file = $cmd{'arg'};
  print "[Tesh/INFO] mkfile $file\n";

  unlink($file);
  open(FILE,">$file") or die "[Tesh/CRITICAL] Unable to create file $file: $!\n";
  print FILE join("\n", @{$cmd{'in'}});
  print FILE "\n" if (scalar @{$cmd{'in'}} > 0);
  close(FILE);
}

# parse tesh file
#my $teshfile=$tesh_file;
#$teshfile=~ s{\.[^.]+$}{};

unless($tesh_file eq "(stdin)"){
  open TESH_FILE, $tesh_file or die "[Tesh/CRITICAL] Unable to open $tesh_file $!\n";
}

my %cmd; # everything about the next command to run
my $line_num=0;
my $finished =0;
LINE: while (not $finished and not $error) {
  my $line;


  if ($tesh_file ne "(stdin)" and !defined($line=<TESH_FILE>)){
    $finished=1;
    next LINE;
  }elsif ($tesh_file eq "(stdin)" and !defined($line=<>)){
    $finished=1;
    next LINE;
  }

  $line_num++;
  chomp $line;
  $line =~ s/\r//g;
  print "[TESH/debug] $line_num: $line\n" if $opts{'debug'};
  my $next;
  # deal with line continuations
  while ($line =~ /^(.*?)\\$/) {
    $next=<TESH_FILE>;
    die "[TESH/CRITICAL] Continued line at end of file\n"
      unless defined($next);
    $line_num++;
    chomp $next;
    print "[TESH/debug] $line_num: $next\n" if $opts{'debug'};
    $line = $1.$next;
  }

  # Push delayed commands on empty lines
  unless ($line =~ m/^(.)(.*)$/) {
    if (defined($cmd{'cmd'}))  {
      exec_cmd(\%cmd);
      %cmd = ();
    }
    print $diff_tool_tmp_fh "$line\n" if ($diff_tool);
    next LINE;
  }

  my ($cmd,$arg) = ($1,$2);
  print $diff_tool_tmp_fh "$line\n" if ($diff_tool and $cmd ne '>');
  $arg =~ s/^ //g;
  $arg =~ s/\r//g;
  $arg =~ s/\\\\/\\/g;
  # handle the commands
  if ($cmd =~ /^#/) {    #comment
  } elsif ($cmd eq '>'){    #expected result line
    print "[TESH/debug] push expected result\n" if $opts{'debug'};
    push @{$cmd{'out'}}, $arg;

  } elsif ($cmd eq '<') {    # provided input
    print "[TESH/debug] push provided input\n" if $opts{'debug'};
    push @{$cmd{'in'}}, $arg;

  } elsif ($cmd eq 'p') {    # comment
    print "[$tesh_name:$line_num] $arg\n";

  } elsif ($cmd eq '$') {  # Command
    # if we have something buffered, run it now
    if (defined($cmd{'cmd'})) {
      exec_cmd(\%cmd);
      %cmd = ();
    }
    if ($arg =~ /^\s*mkfile /){      # "mkfile" command line
      die "[TESH/CRITICAL] Output expected from mkfile command!\n" if scalar @{cmd{'out'}};

      $cmd{'arg'} = $arg;
      $cmd{'arg'} =~ s/\s*mkfile //;
      mkfile_cmd(\%cmd);
      %cmd = ();

    } elsif ($arg =~ /^\s*cd /) {
      die "[TESH/CRITICAL] Input provided to cd command!\n" if scalar @{cmd{'in'}};
      die "[TESH/CRITICAL] Output expected from cd command!\n" if scalar @{cmd{'out'}};

      $arg =~ s/^ *cd //;
      cd_cmd("",$arg);
      %cmd = ();

    } else { # regular command
      $cmd{'cmd'} = $arg;
      $cmd{'file'} = $tesh_file;
      $cmd{'line'} = $line_num;
    }
  }
  elsif($cmd eq '&'){      # parallel command line

    if (defined($cmd{'cmd'})) {
      exec_cmd(\%cmd);
      %cmd = ();
    }
    $cmd{'background'} = 1;
    $cmd{'cmd'} = $arg;
    $cmd{'file'} = $tesh_file;
    $cmd{'line'} = $line_num;
  }
  elsif($line =~ /^!\s*output sort/){    #output sort
    if (defined($cmd{'cmd'})) {
      exec_cmd(\%cmd);
      %cmd = ();
    }
    $cmd{'sort'} = 1;
    if ($line =~ /^!\s*output sort\s+(\d+)/) {
        $sort_prefix = $1;
    }
  }
  elsif($line =~ /^!\s*output ignore/){    #output ignore
    if (defined($cmd{'cmd'})) {
      exec_cmd(\%cmd);
      %cmd = ();
    }
    $cmd{'output ignore'} = 1;
  }
  elsif($line =~ /^!\s*output display/){    #output display
    if (defined($cmd{'cmd'})) {
      exec_cmd(\%cmd);
      %cmd = ();
    }
    $cmd{'output display'} = 1;
  }
  elsif($line =~ /^!\s*expect signal (\w*)/) {#expect signal SIGABRT
    if (defined($cmd{'cmd'})) {
      exec_cmd(\%cmd);
      %cmd = ();
    }
print "hey\n";
    $cmd{'expect'} = "$1";
  }
  elsif($line =~ /^!\s*expect return/){    #expect return
    if (defined($cmd{'cmd'})) {
      exec_cmd(\%cmd);
      %cmd = ();
    }
    $line =~ s/^! expect return //g;
    $line =~ s/\r//g;
    $cmd{'return'} = $line;
  }
  elsif($line =~ /^!\s*setenv/){    #setenv
    if (defined($cmd{'cmd'})) {
      exec_cmd(\%cmd);
      %cmd = ();
    }
    $line =~ s/^! setenv //g;
    $line =~ s/\r//g;
    setenv_cmd($line);
  }
  elsif($line =~ /^!\s*include/){    #include
    if (defined($cmd{'cmd'})) {
      exec_cmd(\%cmd);
      %cmd = ();
    }
    print color("red"), "[Tesh/CRITICAL] need include";
    print color("reset"), "\n";
    die;
  }
  elsif($line =~ /^!\s*timeout/){    #timeout
    if (defined($cmd{'cmd'})) {
      exec_cmd(\%cmd);
      %cmd = ();
    }
    $line =~ s/^! timeout //;
    $line =~ s/\r//g;
    $cmd{'timeout'} = $line;
  } else {
    die "[TESH/CRITICAL] parse error: $line\n";
  }
  if($forked){
   kill(SIGKILL, $forked);
   $timeout=0;
  }

}



# Deal with last command
if (defined($cmd{'cmd'})) {
  exec_cmd(\%cmd);
  %cmd = ();
}


if($forked){
   kill(SIGKILL, $forked);
   $timeout=0;
}

foreach(@bg_cmds){
  my %test=%{$_};
  waitpid ($test{'pid'}, 0);
  $test{'gotret'} = exit_status($?);
  parse_out(\%test);
}

@bg_cmds=();

if ($diff_tool) {
  close $diff_tool_tmp_fh;
  system("$diff_tool $diff_tool_tmp_filename $tesh_file");
  unlink $diff_tool_tmp_filename;
}

if($error !=0){
    exit $exitcode;
}elsif($tesh_file eq "(stdin)"){
    print "Test suite from stdin OK\n";
}else{
    print "Test suite `$tesh_name' OK\n";
}

#my (@a,@b);
#push @a,"bl1";   push @b,"bl1";
#push @a,"bl2";   push @b,"bl2";
#push @a,"bl3";   push @b,"bl3";
#push @a,"bl4";   push @b,"bl4";
#push @a,"bl5";   push @b,"bl5";
#push @a,"bl6";   push @b,"bl6";
#push @a,"bl7";   push @b,"bl7";
##push @a,"Perl";  push @b,"ruby";
#push @a,"END1";   push @b,"END1";
#push @a,"END2";   push @b,"END2";
#push @a,"END3";   push @b,"END3";
#push @a,"END4";   push @b,"END4";
#push @a,"END5";   push @b,"END5";
#push @a,"END6";   push @b,"END6";
#push @a,"END7";   push @b,"END7";
#print "Identical:\n". build_diff(\@a,\@b);

#@a = (); @b =();
#push @a,"AZE"; push @b,"EZA";
#print "Different:\n".build_diff(\@a,\@b);

use lib "@CMAKE_BINARY_DIR@/bin" ;

use Diff qw(diff); # postpone a bit to have time to change INC

sub build_diff {
  my $res;
  my $diff = Diff->new(@_);

  $diff->Base( 1 );   # Return line numbers, not indices
  my $chunk_count = $diff->Next(-1); # Compute the amount of chuncks
  return ""   if ($chunk_count == 1 && $diff->Same());
  $diff->Reset();
  while(  $diff->Next()  ) {
    my @same = $diff->Same();
    if ($diff->Same() ) {
      if ($diff->Next(0) > 1) { # not first chunk: print 2 first lines
        $res .= '  '.$same[0]."\n" ;
        $res .= '  '.$same[1]."\n" if (scalar @same>1);
      }
      $res .= "...\n"  if (scalar @same>2);
#    $res .= $diff->Next(0)."/$chunk_count\n";
      if ($diff->Next(0) < $chunk_count) { # not last chunk: print 2 last lines
        $res .= '  '.$same[scalar @same -2]."\n" if (scalar @same>1);
        $res .= '  '.$same[scalar @same -1]."\n";
      }
    }
    next if  $diff->Same();
    map { $res .= "- $_\n" } $diff->Items(1);
    map { $res .= "+ $_\n" } $diff->Items(2);
  }
  return $res;
}


